File: languages/c/fannkuch.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static void fannkuch(int n, int *max_flips_out, int *checksum_out) {
    int *perm1 = malloc(n * sizeof(int));
    int *perm = malloc(n * sizeof(int));
    int *count = malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) perm1[i] = i;

    int max_flips = 0, checksum = 0, r = n;
    long perm_idx = 0;  /* counts up to n!-1 → needs 64-bit for n >= 13 */
    for (;;) {
        while (r != 1) { count[r - 1] = r; r--; }

        memcpy(perm, perm1, n * sizeof(int));
        int flips = 0, k;
        while ((k = perm[0]) != 0) {
            for (int i = 0, j = k; i < j; i++, j--) {
                int t = perm[i]; perm[i] = perm[j]; perm[j] = t;
            }
            flips++;
        }

        if (flips > max_flips) max_flips = flips;
        checksum += (perm_idx % 2 == 0) ? flips : -flips;

        /* Generate the next permutation. */
        for (;;) {
            if (r == n) {
                *max_flips_out = max_flips; *checksum_out = checksum;
                free(perm1); free(perm); free(count);
                return;
            }
            int first = perm1[0];
            for (int i = 0; i < r; i++) perm1[i] = perm1[i + 1];
            perm1[r] = first;
            if (--count[r] > 0) break;
            r++;
        }
        perm_idx++;
    }
}

int main(int argc, char **argv) {
    int n = argc > 1 ? atoi(argv[1]) : 7;
    int max_flips, checksum;
    fannkuch(n, &max_flips, &checksum);
    printf("%d\n", checksum);
    printf("Pfannkuchen(%d) = %d\n", n, max_flips);
    return 0;
}

File: languages/ruby/fannkuch.rb
def fannkuch(n)
  perm1 = (0...n).to_a
  count = Array.new(n, 0)
  max_flips = 0
  checksum = 0
  perm_idx = 0
  r = n
  loop do
    while r != 1
      count[r - 1] = r
      r -= 1
    end

    perm = perm1.dup
    flips = 0
    k = perm[0]
    while k != 0
      i = 0
      j = k
      while i < j
        perm[i], perm[j] = perm[j], perm[i]
        i += 1
        j -= 1
      end
      flips += 1
      k = perm[0]
    end

    max_flips = flips if flips > max_flips
    checksum += perm_idx % 2 == 0 ? flips : -flips

    # Generate the next permutation.
    loop do
      return [max_flips, checksum] if r == n
      first = perm1[0]
      i = 0
      while i < r
        perm1[i] = perm1[i + 1]
        i += 1
      end
      perm1[r] = first
      count[r] -= 1
      break if count[r] > 0
      r += 1
    end
    perm_idx += 1
  end
end

n = ARGV[0] ? ARGV[0].to_i : 9
max_flips, checksum = fannkuch(n)
puts checksum
puts "Pfannkuchen(#{n}) = #{max_flips}"
